### ISSUE TITLE
Implement 2FA badge on user profiles

### ISSUE THREAD

**Issue by todb** [mastodon/mastodon#21653]
### Pitch

Users who have opted into 2FA should get a special badge on their profile that indicates they've set up 2FA. A lock or something. I like GitHub's [two-keyhole lock design](https://user-images.githubusercontent.com/24144/204069767-781362ff-f1dc-4021-8c08-96c1608fc375.png), but even a regular emoji would do the trick.

This badge should show up in a region the user normally can't edit, to avoid stolen 2FA valor (much like how people put :verified: badge emojis in their username or link lists).

2FA status should be encoded in a user profile API, so other clients can do what they want with this information.

If you're feeling particularly pushy, users that don't have 2FA enabled could also have a badge saying as such. This would help solve the screen real estate problem if there's a region dedicated to is/is not 2FA.

### Motivation

2FA is a bit of a hassle to set up and use, so users who choose to secure their accounts with it should be rewarded -- MMORPG's show us that people absolutely love collecting badges.

Also, advertising that your account has 2FA enabled makes it easier and wiser to trust that the account is not currently compromised, especially for famous / notable people. It also helps normalize the behavior of setting up 2FA.

2FA, finally, helps combat spam signups (see #13013), so making it more obvious that it's a feature people can use might help uptake, which in turn might make moving to a mandatory 2FA requirement easier for instances that want that.

**Comment by quardbreak** [comment]
>2FA is a bit of a hassle to set up and use, so users who choose to secure their accounts with it should be rewarded -- MMORPG's show us that people absolutely love collecting badges.

But this is not a game. People use 2FA because they wish more security, not because this is a quest or a community contest where you can achieve something. This is not something to "brag about". Sensitive settings must be viewed only by a person who set them, otherwise they just reveals everyone your security status, which is, IMHO, an awful approach to security design.

**Comment by robsonsobral** [comment]
I agree with your suggestions, but disagree with one of your reasons. As @quardbreak said, it isn't a game. However, it would be a good way to promote a good practice.

**Comment by eishiya** [comment]
While it's good to promote good security practices, I don't think making your 2FA status public and essentially shaming those who are unwilling or unable to use 2FA is a good idea, especially since it also increases their chance of being targeted by letting bad actors see who is less secure.

I think a better approach would be highlighting that menu item for users who don't have it set up, and showing a reminder to add 2FA when they log in. That way, they'll be gently nagged about it at appropriate times (when they're logging in, and when they're looking at settings anyway), but in a way that doesn't reveal their lack of 2FA to people who would act on that status.

**Comment by todb-r7** [comment]
On the point of using an incentive like a badge: recall when Epic Games associated 2FA with a special collectible dance move in Fortnite. It was a huge innovation that I bet drove 2FA adoption among completists. More [here](https://www.epicgames.com/help/en-US/epic-accounts-c5719348850459/account-security-c5719366891291/what-rewards-do-i-get-for-enabling-two-factor-authentication-2fa-on-my-epic-games-account-a5720233583259). Yes, Fortnite is a game, but Mastodon can be fun, too.

On targeting non-2FA users: this is true! Implementation would mean sticking a "Hack Me!" sign on the backs of everyone who doesn't have 2FA. Callously, though, wouldn't this drive adoption of 2FA, though? Right now, the people who are vulnerable to account takeover are indistinguishable from those who aren't - for regular people without the technical means to test for it. I feel like this kind of partial information situation benefits attackers.

**Comment by eishiya** [comment]
>I feel like this kind of partial information situation benefits attackers.

Even if 2FA were a reliable indicator of "this user has not been hacked", which it is not, I don't think it's worth the other drawbacks. I'm against needless gamification in general - it's fun for many users, but unpleasant or even harmful for others. Mastodon is already fun. 2FA doesn't need to be "fun", it needs to be easy to find and set up.

I think people should be made aware that 2FA is available and educated about its benefits, and choose for themselves whether it's appropriate for them, rather than being incentivised (=pressured or guilted, for those negatively affected by public indicators) through unrelated means.

**Comment by blaine** [comment]
There are many reasons why people might not activate 2FA.

SMS-based 2FA, for example, isn't usable by people who don't have cell reception at home (as is the case for most people in my rural community), or who don't have a reliable phone plan for any reason. Similarly, even for Mastodon's TOTP-based 2FA, for people whose accounts aren't at risk, having 2FA enabled and losing or not having access to the authenticator (e.g., new/lost phone, accidentally deleting app, shared or public computer, etc) is worse than not having 2FA in the first place.

Also, advertising non-2FA status would put a target on all users who chose to not use it.

Better would be to build more robust authentication flows, including adding Passkey/WebAuthn support and OIDC SSO support for instances where that's appropriate. e.g., Using OIDC for supported domains versus account confirmation / password reset emails would improve security much more than 2FA support for a huge number of users (due to not depending on weak passwords), and Passkey support would do the same for most users  with the added benefit of not requiring instance admins to set up dev accounts with Google et al.
